Source/Global/perform_env.cpp

```diff
@@ -210,32 +208,130 @@ Result<HC_UNIQUE_PTR<HC_PERFORM_ENV>> HC_PERFORM_ENV::Initialize(HCInitArgs* arg
 #elif HC_PLATFORM == HC_PLATFORM_ANDROID
     auto initAndroidResult = AndroidPlatformContext::Initialize(args);
     RETURN_IF_FAILED(initAndroidResult.hr);
-
-    performEnv.reset(new (a.allocate(1)) HC_PERFORM_ENV);
     performEnv->androidPlatformContext = initAndroidResult.ExtractPayload();
 
 #else
-    UNREFERENCED_PARAMETER(a);
     RETURN_HR_IF(E_INVALIDARG, args);
 #endif
 
     return std::move(performEnv);
 }
 
+struct HC_PERFORM_ENV::HttpPerformContext
+{
+    HttpPerformContext(HC_PERFORM_ENV* _env, HCCallHandle _callHandle, XAsyncBlock* _clientAsyncBlock) :
+        env{ _env },
+        callHandle{ _callHandle },
+        clientAsyncBlock{ _clientAsyncBlock },
+        internalAsyncBlock{ nullptr, this, HC_PERFORM_ENV::HttpPerformComplete }
+    {
+    }
+
+    ~HttpPerformContext()
+    {
+        XTaskQueueCloseHandle(internalAsyncBlock.queue);
+    }
+
+    HC_PERFORM_ENV* env{};
+    HCCallHandle const callHandle{};
+    XAsyncBlock* const clientAsyncBlock{};
+    XAsyncBlock internalAsyncBlock{};
+};
+
+HRESULT HC_PERFORM_ENV::HttpCallPerformAsyncShim(HCCallHandle call, XAsyncBlock* async)
+{
+    auto performContext = http_allocate_shared<HttpPerformContext>(this, call, async);
+    {
+        std::unique_lock<std::mutex> lock{ m_mutex };
+        m_activeHttpRequests[performContext->callHandle] = performContext;
+    }
+    return XAsyncBegin(async, performContext.get(), nullptr, __FUNCTION__, HttpPerformAsyncShimProvider);
+}
+
 HRESULT HC_PERFORM_ENV::CleanupAsync(HC_UNIQUE_PTR<HC_PERFORM_ENV>&& env, XAsyncBlock* async) noexcept
 {
     RETURN_IF_FAILED(XAsyncBegin(async, env.get(), __FUNCTION__, __FUNCTION__, CleanupAsyncProvider));
     env.release();
     return S_OK;
 }
 
+HRESULT CALLBACK HC_PERFORM_ENV::HttpPerformAsyncShimProvider(XAsyncOp op, const XAsyncProviderData* data)
+{
+    HttpPerformContext* performContext{ static_cast<HttpPerformContext*>(data->context) };
+    HC_PERFORM_ENV* env{ performContext->env };
+
+    switch (op)
+    {
+    case XAsyncOp::Begin:
+    {
+        XTaskQueuePortHandle workPort{};
+        RETURN_IF_FAILED(XTaskQueueGetPort(data->async->queue, XTaskQueuePort::Work, &workPort));
+        RETURN_IF_FAILED(XTaskQueueCreateComposite(workPort, workPort, &performContext->internalAsyncBlock.queue));
+
+        return performContext->callHandle->PerformAsync(&performContext->internalAsyncBlock);
+    }
+    case XAsyncOp::Cancel:
+    {
+        XAsyncCancel(&performContext->internalAsyncBlock);
+        return S_OK;
+    }
+    case XAsyncOp::Cleanup:
+    {
+        std::unique_lock<std::mutex> lock{ env->m_mutex };
+        env->m_activeHttpRequests.erase(performContext->callHandle);
+        bool scheduleCleanup = env->m_activeHttpRequests.empty() && env->m_cleanupAsyncBlock;
+        lock.unlock();
+
+        if (scheduleCleanup)
+        {
+            HRESULT hr = XAsyncSchedule(env->m_cleanupAsyncBlock, 0);
+            assert(SUCCEEDED(hr)); // How to handle failure here?
```

Comment:
normally I would just log and ignore cleanup errors

Reply:
Leaving as assert for now.  Need to think a bit more about why this could fail and how we should proceed if it does.